api: Implement concurrent tasks rate-limiting for VOD

victorges opened the pull request:
**What does this pull request do? Explain your changes. (required)**
This is to implement a couple rate-limiting protections on the VOD pipeline against exaggerated traffic
from specific users. There are 3 main and independent strategies for rate-limiting implemented here.

**Specific updates (required)**
The strategies are:
1.   A maximum number of tasks that the pipeline can run for the same user concurrently.
  - This works by querying for tasks in the `running` or "retrying" state, and returning 429 on the initial request made by `task-runner` to start executing the task. Task runner learned how to process that here https://github.com/livepeer/task-runner/pull/95 and will backoff the message on the delayed queue for a minute, before trying to start again
2. A maximum number of tasks that can be enqueued for the same user. This includes the tasks that are currently executing.
  - Differently from the previous strategy, this doesn't just backoff the execution of the task, but it returns an explicit error to the user when they try to create more tasks. 
  - The point here is to protect the delayed queue which is used in the first strategy. We delay events for 1 minute at each step, so there is a number of tasks being delayed at which we would not be able to process even the quick-delayed-tasks in a timely manner. Assuming a latency of 100ms to process each message and the same 15x concurrency on a region/queue, this would mean that `60*1000/100*15=9000` messages could create a problem if we didn't have this rate limiting, which seems pretty possible.
  - Based on the numbers above, setup a limit of 100 tasks (by default, everything is configurable) enqueued for each user.
3. A deduplication strategy to avoid processing the same file multiple times. Yeah this might be annoying for testing, but will protect us in production from possible bugs in client implementations or our own (like the old livepeer.js bug that published created of assets for the same URL). The neat thing here is that we don't return an error: we just return the already running/duplicate task back.

## -

- **How did you test each of these updates (required)**
`yarn test`, transcoded dozens of files in staging, tried uploading 200 files at once (got rate limited), observed how execution is now happening 5 tasks at a time, pretty orderly, and tried publishing duplicate assets and ensured same asset/task was returned.

**Does this pull request close any open issues?**
I think so but idk where it is. It's a blocker for MediaCovnert launch on 12/16.

**Checklist:**

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have read the **CONTRIBUTING** document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

victorges committed:
api/db: Create query to fetch recent assets with same source URL

victorges committed:
api/asset: Prohibit duplicate asset uploads

victorges committed:
api/asset: Create test for duplicate detection

victorges committed:
api/task: Create rate limit in number of running tasks

victorges committed:
api/task: Allow admins to bypass concurrent limit

Record tester needs the parallelism

victorges committed:
api: Fix trailing parenthesis

victorges committed:
api/task: Fix the task we fail

victorges committed:
api/task: Fix query for running tasks

victorges committed:
api/task: Filter lost tasks directly in the query

victorges committed:
api: Fix duplicate asset query

victorges committed:
api: Cleanup asset dup logic

victorges committed:
api/task: Move new query to proper task table class

victorges committed:
api/asset: Fix call to db to find duplicate upload

victorges committed:
api: Create query for enqueued tasks

victorges committed:
api: Add some sugar on top of task query

victorges committed:
api: Add limit on maximum enqueued tasks per user

victorges committed:
api: Remove ORDER BY from query

victorges committed:
Add debug log

victorges committed:
api: Avoid double parenthesis tuple syntax

victorges committed:
api: Fix concurrent tasks query

victorges committed:
api: Make all AMQP messages mandatory

victorges committed:
api/asset: Return correct asset on duplicate

victorges committed:
api/queue: Add TODO about handling AMQP returns

victorges committed:
api/store: Lower enqueued task timeout

thomshutt commented on packages/api/src/store/asset-table.ts:
```diff
@@ -72,4 +73,21 @@ export default class AssetTable extends Table<WithID<Asset>> {
+    url: string,
+    userId: string
+  ): Promise<WithID<Asset>> {
+    const createdAfter = Date.now() - DUPLICATE_ASSETS_THRESHOLD;
```
Good idea, I think DUPLICATE_ASSETS_THRESHOLD saves us from most weird states (something ends up permanently in `waiting` / `processing` etc.)

victorges committed:
api: s/concurrent/pending

victorges committed:
Revert "Add debug log"

This reverts commit 32cb297bd5fc4956815a83a005a7c934c0a8f2d1.

victorges committed:
api/queue: Remove the mandatory stuff that we don't handle

victorges committed:
api/store: Create tasks index for pending tasks query

victorges committed:
api/store: Use a regular template string for CREATE INDEX

Psql doesn't like the syntax when we use the template strings,
I guess there's no prepared statements for schema modifying queries.
